pkg/storegateway/series_chunks_test.go

```diff
@@ -884,31 +884,49 @@ func (s *sliceSeriesChunksSetIterator) Err() error {
 	return nil
 }
 
-// delayedSetIterator implements iterator and
-// introduces an artificial delay before returning from Next() and At().
-type delayedSetIterator[S any] struct {
+// delayedIterator implements iterator and introduces an artificial delay before returning from Next() and At().
+type delayedIterator[S any] struct {
 	wrapped iterator[S]
-	delay   time.Duration
+
+	chance int // a chance for the delay to happen
+	delay  time.Duration
+}
+
+func newDelayedIterator[S any](delay time.Duration, wrapped iterator[S]) *delayedIterator[S] {
+	return &delayedIterator[S]{
+		wrapped: wrapped,
+		delay:   delay,
+	}
 }
 
-func newDelayedSetIterator[S any](delay time.Duration, wrapped iterator[S]) *delayedSetIterator[S] {
-	return &delayedSetIterator[S]{
+func newDelayedMaybeIterator[S any](chance int, delay time.Duration, wrapped iterator[S]) *delayedIterator[S] {
+	if chance < 0 || chance > 100 {
+		panic("delayedIterator: probability must be between 0 and 100")
+	}
+	return &delayedIterator[S]{
 		wrapped: wrapped,
+		chance:  chance,
 		delay:   delay,
 	}
 }
 
-func (s *delayedSetIterator[S]) Next() bool {
-	time.Sleep(s.delay)
+func (s *delayedIterator[S]) delayMaybe() {
+	if s.chance == 0 || rand.Intn(100) <= s.chance {
+		time.Sleep(s.delay)
```

Comment:
wouldn't this randomness make the benchmarks less reproducible/consistent?

Reply:
the total run of these benchmarks looks to be at most 10ms. I think missing one or a few of these `time.Sleep` can make a significant difference (I consider +/- 10% to be significant).

Would it be better to set the delay to be constant on all `Next()` invocations, but reduce the duration? So if we are aiming for 1ms sleep on 5% of invocations, then this is statistically the same time slept as 50µs sleep on every invocation. In this case on average the time to sleep is the same in both cases, but the average, best, and worst cases are the same when we sleep a consistent amount on every invocation.

Reply:
As discussed in DM, my initial idea was to avoid having IO (represented by a `sleep`) on every call to underlying's `Iterator.Next()`. This shouldn't have been a goal in the first place, and it only makes it confusing.

I update this chunk to use a fixed delay. The results of these benchmarks are still the same (updated the description of the PR, also).